core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -258,7 +258,7 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
     RelDataType fromType = validator.deriveType(scope, node);
     // This depends on the fact that type validate happens before coercion.
     // We do not have inferred type for some node, i.e. LOCALTIME.
-    if (fromType == null) {
+    if (fromType == null || fromType.getSqlTypeName() == SqlTypeName.NULL) {
```

Comment:
I don't see why this is right, and I don't see why it's needed for this optimization.
In general many tests will fail if you supply a NULL, but will pass if you cast the NULL to the correct type. So I suspect this is actually needed.

Reply:
@mihaibudiu Yes, I will check again. After this PR, some unit test behavior is unexpected.